opentelemetry-appender-tracing/src/layer.rs

```diff
@@ -1,60 +1,122 @@
-use opentelemetry::logs::{LogRecord, Logger, LoggerProvider, Severity};
+use opentelemetry::{
+    logs::{AnyValue, LogRecord, Logger, LoggerProvider, Severity},
+    Key,
+};
 use std::borrow::Cow;
-use tracing_core::Level;
+use tracing_core::{Level, Metadata};
+use tracing_log::NormalizeEvent;
 use tracing_subscriber::Layer;
 
 const INSTRUMENTATION_LIBRARY_NAME: &str = "opentelemetry-appender-tracing";
+// Start by "log."
+const LOG_METADATA_ATTRIBUTES: &[&str] = &[
+    "name",
+    "target",
+    "module.path",
+    "file.path",
+    "file.name",
+    "file.line",
+    "module_path",
+    "file",
+    "line",
+];
 
 /// Visitor to record the fields from the event record.
-struct EventVisitor<'a> {
-    log_record: &'a mut LogRecord,
+#[derive(Default)]
+struct EventVisitor {
+    log_record_attributes: Vec<(Key, AnyValue)>,
+    log_record_body: Option<AnyValue>,
 }
 
-impl<'a> tracing::field::Visit for EventVisitor<'a> {
+fn is_metadata(field: &str) -> bool {
+    if let Some(log_field) = field.strip_prefix("log.") {
+        for log_metadata_field in LOG_METADATA_ATTRIBUTES {
```

Comment:
Rather than a `for` loop to search on string compares of every entry, I think a `match` on the set of literals in your `LOG_METADATA_ATTRIBUTES` might generate better code, as I could see the compiler sorting the branch check order to take advantage of bailing out early and avoiding having to check all entries.  e.g.

``` rust
fn is_metadata(field: &'static str) -> bool {
    field
        .strip_prefix("log.")
        .map(|remainder| {
            matches!(
                remainder,
                    "file.line"
                    | "file.name"
                    | "file.path"
                    | "file"
                    | "line"
                    | "module_path"
                    | "module.path"
                    | "name"
                    | "target"
            )
        })
        .unwrap_or(false)
}
```
(I've manually sorted the list of matches, but I'm not sure if its required for better code or not.)
This is also a little more concise than the current version, fwiw.

Reply:
Thanks! I will test it and push it.